rfcs/text/0014_notification_service.md

```diff
@@ -0,0 +1,373 @@
+- `markAsUnread({ notification_id: string[] })` - mark notifications with given ids as unread.
+- `markAllAsRead()` - mark all the notifications as read.
+- `markAsPin({ notification_id: string[] })` - mark notifications with given ids as pinned.
+- `markAsUnpin({ notification_id: string[] })` - mark notifications with given ids as unpinned.
+- `configure()` - updates user-specific notification settings.
+
+#### LocalRepository
+![image](../images/0014/local_repository.png)
+
+##### Notification model
+Notification model is responsible for:
+- Applying policies to an incoming notification. NS must check whether a recipient unsubscribed from a source before processing to delivery.
+- Enhancing an incoming notification with necessary data: 
+    - `expired_at` if not specified. Kibana allows administrators to set how long notifications will be stored. Falls back to 30 - 90 days be default. 
+- Writing new incoming messages to the Notification storage.
+
+Processing a large number of incoming notifications might slow down Kibana and create a disproportionately high load on some Kibana instances.
+The use of a queue allows us to make the handling of incoming notifications asynchronously.
+Async mechanism lets Kibana to cope with spikes in the number of messages created. The queue implementation is based on the Elasticsearch server. 
+
+The logic is run as a background task. The usage of the single background task:
+- reduces the possibility of read- and write-conflicts
+- allows re-using task_manager built-in features: scheduling, retry, monitoring
+
+If Kibana processes notifications in asynchronous batches, we will need a prioritization mechanism to be sure events
+with a higher severity are delivered first without a significant delay. The prioritization mechanism is based on
+priority property value: the notification with the higher value is processed earlier to provide an experience as close to real-time as possible.
+
+##### Notification storage
+
+NS storage keeps different notification statuses for every user. Most Kibana deployments have [from 20 to 90 users](https://elastic.slack.com/archives/C9097ABGC/p1597162974399900),
+although there are some odd instances with >100k users. So we consider two options at the moment (see the [workflow](https://docs.google.com/document/d/18s-BTHog_oPaQKf871gzuhxkah4pK1GYJM9DCLceaLo/edit#)):
+- Store notification state separately from a notification. State for new notifications created for every user when they change the state.
+```typescript
+interface  NotificationState {
+  notification_id: string;
+  recipient_id: string;
+  read_status: boolean;
+  pinned_status: boolean;
+}
+``` 
+This option consumes less storage space but might lead to expensive read operations due to an additional JOIN call to 
+search for notification state objects. We don’t expect Kibana to read more than 10-20 notifications at once, so the overhead might be acceptable.
```

Comment:
I don't believe this extra query would be very expensive, but it would require one additional round trip for each page of notifications. That seems pretty acceptable to me.

If we have the same, monotonically increasing sort key on both the notifications and the state documents, I believe we could do these queries in parallel on the primary notification view, which is chronological. This is possible because the last X notifications and last X state documents should match up.

The challenge is that Elasticsearch doesn't give us a way to create monotonically increasing IDs, so we'd most likely need to use timestamps + falling back to 2 serial queries for issues where the clock is not synced. But this is definitely a performance optimization that we would only need to do if necessary.